<a href="https://colab.research.google.com/github/TMPenwell/MIS2800/blob/main/MIS2800_PRACTICE_EXAM_1_ANSWER_KEY.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# IMPORTANT ❗
The below blocks are the lines of code to begin using sqlite3 magic.  

This is meant to be reused as needed.  To use, go to File > Save a Copy in Drive.


In [1]:
!pip install --upgrade pandas ipython-sql prettytable==3.10.1


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 79.5/79.5 kB 2.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.9/10.9 MB 92.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 78.3 MB/s eta 0:00:00
  Attempting uninstall: prettytable
    Found existing installation: prettytable 3.17.0
    Uninstalling prettytable-3.17.0:
      Successfully uninstalled prettytable-3.17.0
  Attempting uninstall: pandas
    Found existing installation: pandas 2.2.2
    Uninstalling pandas-2.2.2:
      Successfully uninstalled pandas-2.2.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 3.0.1 which is incompatible.
gradio 5.50.0 requires pandas<3.0,>=1.0, but you have pandas 3.0.1 which is incompatible.
db-dtypes 1.5.0 requires pandas<3.0.0,>=1.5.3, but you have pandas 3.0.1 which i

In [2]:
# Install the SQL extension. Remember to type exactly as shown here.
!pip install ipython-sql
print("ipython sql installed")

ipython sql installed


In [3]:
# Load the SQL magic extension
%load_ext sql

In [4]:
#this code imports sqlite3 and prints a statement stating it has been imported
import sqlite3
print("sqlite3 imported")

sqlite3 imported


In [5]:
# This creates a brand-new, empty database file named [name of your database.db]
conn = sqlite3.connect('nameOfDatabase.db')
print("nameOfDatabase database created")


nameOfDatabase database created


In [6]:
# This line is required so Colab stops acting like Python and acts more like a SQL database terminal
%sql sqlite:///nameOfDatabase.db

# ----SPECIFIC FOR PREP/IC/HW ----

In [ ]:
%%sql
CREATE TABLE Events (
    event_id INTEGER PRIMARY KEY,
    event_name TEXT,
    sport TEXT,
    event_date TEXT,
    venue TEXT
) STRICT;
-- sqllite does not support DATE datatype - using TEXT instead

In [16]:
# confirm the table was created
# NOTE: this was not part of your exercise
%%sql
SELECT name FROM sqlite_master WHERE type='table';


 * sqlite:///nameOfDatabase.db
Done.


name
Events


In [18]:
%%sql
INSERT INTO Events (event_id, event_name, sport, event_date, venue)
VALUES
    (1, '100m Sprint', 'Track', '2028-07-21', 'Main Stadium'),
    (2, 'Swimming 200m Freestyle', 'Swimming', '2028-07-23', 'Aquatics Center'),
    (3, 'Gymnastics All-Around', 'Gymnastics', '2028-07-25', 'Arena Complex');

 * sqlite:///nameOfDatabase.db
3 rows affected.


[]

In [19]:
%%sql
SELECT * FROM Events
WHERE sport = 'Track';

 * sqlite:///nameOfDatabase.db
Done.


event_id,event_name,sport,event_date,venue
1,100m Sprint,Track,2028-07-21,Main Stadium


In [20]:
%%sql
UPDATE Events
SET venue = 'Olympic Stadium'
WHERE event_id = 1;

 * sqlite:///nameOfDatabase.db
1 rows affected.


[]

# Multiple Choice

# 1. OLTP systems are primarily designed for:

A. Historical trend analysis
B. **Real-time transactions**
C. Data warehousing
D. Reporting dashboards

***Online Level Transaction Processing (OLTP) is the "front-line" system. It is optimized for speed and accuracy in handling day-to-day business operations—like a customer buying a jersey or an athlete registering for a race. It focuses on CRUD (Create, Read, Update, Delete) in the moment.***

# 2. Which of the following is characteristic of OLAP?

A. Many small insert operations
B. Frequent updates
C. **Complex aggregate queries**
D. High transaction speed requirements


***Online Analytical Processing (OLAP) is the "big picture" system. It isn't used for single sales; it’s used to analyze millions of rows at once to find trends. These systems handle "aggregates" like SUM, AVG, and COUNT over long periods of history.***

# 3. In a star schema, the center table is called:

A. Dimension table
B. Lookup table
C. **Fact table**
D. Bridge table



***The center table of a star schema is the Fact Table. It acts as the central hub, storing quantitative, numerical, and measurable data (measures or metrics) such as sales amounts, quantities, or profit. This table is surrounded by dimension tables and typically contains foreign keys that link to them.***

**Analogy: *Think of the Fact Table as the box score of a game. It contains the measurable, quantitative data (the "facts" or numbers, like points scored or items sold). The surrounding Dimension Tables (the points of the star) contain the descriptive data like names, dates, and locations.***

# 4. First Normal Form (1NF) requires:

A. No foreign keys
B. **No repeating groups**
C. No primary keys
D. Only one table



***1NF is about "Atomicity," meaning data cannot be broken down any further. If a cell contains a list (like "QB, Captain"), it violates 1NF. Every cell must contain exactly one piece of information—one box, one jersey.***

Just as an atom was historically considered the smallest unit of matter that couldn't be split without losing its identity, an atomic value in a database is the smallest unit of data that should be stored in a single cell.

In science, if you split an atom, you’re no longer dealing with the same element. In SQL, if you "split" a truly atomic value (like the number 7), you just have meaningless fragments.

A cell containing "QB, Captain" is like a molecule—it's a compound made of two distinct parts. 1NF requires us to break that molecule down into its individual atoms.

"Think of 1NF as the 'Atomic Phase.' Just as an atom is the most singular form of matter, an atomic value is the most singular form of data.

If a cell contains a list (like 'QB, Captain'), it’s a molecule, not an atom. To reach 1NF, we must break it down until every cell holds exactly one 'atom' of information.

Note: What is 'atomic' depends on your goal. A full name 'Joe Burrow' might be an atom in a simple list, but in a professional database, the atoms are actually 'Joe' and 'Burrow.'"


# Molecule (Non-Atomic): A cell containing "Joe Burrow, QB, LSU"



# Atoms (1NF): Three separate columns for Name, Position, and College.




# 5. Which data type should be used for medal count?

A. TEXT
B. **INTEGER**
C. DATE
D. BOOLEAN


***An INTEGER is a whole number. Since you cannot win 1.5 gold medals, you don't need a REAL (decimal) type. You also can't perform math on TEXT, so INTEGER is the only logical choice for counting.***

# 6. A Foreign Key:

A. Must be unique
B. Identifies each row uniquely
C. **Links to a Primary Key in another table**
D. Cannot be NULL


***A Foreign Key is how we create a relationship between two tables. It is a column in one table that points back to the Primary Key in another. It’s the "Coach’s Clipboard" referencing the "Player’s Jersey Number."***